client/starwhale/api/_impl/data_store.py

```diff
@@ -127,12 +137,30 @@ def decode_schema(schema: Dict[str, Any]) -> "SwType":
             element_type = schema.get("elementType", None)
             if element_type is None:
                 raise RuntimeError(f"no element type found for type {type_name}")
-            return SwListType(SwType.decode_schema(element_type))
+            attrs = schema.get("attributes", None)
+            sparse_types: Dict[int, SwType] | None = None
+            if attrs is not None:
+                if not isinstance(attrs, list):
+                    raise RuntimeError("attributes should be a list")
+                sparse_types = {}
+                for attr in attrs:
+                    index = attr.get("index", None)
+                    if index is None:
+                        raise RuntimeError("no index found for attribute")
+                    sparse_types[index] = SwType.decode_schema(attr)
+            return SwListType(
+                element_types=SwType.decode_schema(element_type),
+                sparse_types=sparse_types,
+            )
         if type_name == "TUPLE":
-            element_type = schema.get("elementType", None)
-            if element_type is None:
-                raise RuntimeError(f"no element type found for type {type_name}")
-            return SwTupleType(SwType.decode_schema(element_type))
+            element_types = schema.get("elementTypes", None)
+            if element_types is None:
+                element_type = schema.get("elementType", None)
+                if element_type is None:
+                    raise RuntimeError(f"no element type found for type {type_name}")
+                element_types = [element_type]
+            decoded_element_types = [SwType.decode_schema(t) for t in element_types]
+            return SwTupleType(decoded_element_types)
```

Comment:
Should there get the attrs and pass it to SwTupleType constructor?

Reply:
Done